### ISSUE TITLE
missing-field-initializers warning not reporting missing designated initializers

### ISSUE THREAD

**Issue by ottojo** [llvm/llvm-project#56628]
In the following code, the `missing-field-initializers` does not produce a warning that `b` is not explicitly initialized.

Changing `CheckForMissingFields` to `True` here:
https://github.com/llvm/llvm-project/blob/a9a60f20e6cc80855864b8f559073bc31f34554b/clang/lib/Sema/SemaInit.cpp#L2167-L2169
seems to produce the expected warning.

The comment suggests
```cpp
// Disable check for missing fields when designators are used.
// This matches gcc behaviour.
```
However, GCC *does* produce the warning in that case.
Comparison with GCC: https://godbolt.org/z/WeYdY11q3

```c++
struct ExampleStruct {
    int a;
    int b;
};

ExampleStruct buildExampleStruct(int a) {
    ExampleStruct e{
            .a=a
    };
    return e;
}
```

Build flags: `-Werror=missing-field-initializers -std=c++20`

Expected behavior:
```console
main.cpp:10:5: error: missing field 'b' initializer [-Werror,-Wmissing-field-initializers]
```
matching GCCs
```console
<source>:9:5: error: missing initializer for member 'ExampleStruct::b' [-Werror=missing-field-initializers]
```

Actual behavior: Compiles without warning/error.

**Comment by dbelousoff** [comment]
Is there any updates/workarounds for this?

**Comment by shafik** [comment]
I don't see a workaround, it looks out of date at this point. Someone needs to remove the `CheckForMissingFields = false;` but this is a potentially breaking change since there may be folks that using this flag that won't expect the warning to start popping up. Although it should.

**Comment by ottojo** [comment]
I see how this would be a breaking change, although i guess if someone enabled Wextra and Wall they would expect some breakage when updating the compiler...

Anyways, i felt like hacking at this a bit and maybe introducing a dedicated switch for this might be an idea? I think the warning is useful, and i would certainly enable it in my projects (although discovery of this flag might be a problem...)
An attempt of an implementation is here, which seems to work in the above test at least :smile: https://github.com/ottojo/llvm-project/pull/2/files

**Comment by gracicot** [comment]
I'm actually hit by this issue and I actually missed a bug in my code because I assumed the warning was working on Clang too.

**Comment by shafik** [comment]
@AaronBallman sounds like missing this case is not ideal, wdyt about making this change?

**Comment by AaronBallman** [comment]
> @AaronBallman sounds like missing this case is not ideal, wdyt about making this change?

I'm comfortable with the change. If it turns out to be highly disruptive, we can always put the diagnostic under a new warning flag grouped under `-Wmissing-field-initializers` so that users can selectively disable this (but hopefully we won't need such a measure).

**Comment by Fznamznon** [comment]
https://reviews.llvm.org/D157879

**Comment by jamesruan** [comment]
It is how we understand designated initializers that matters.
IMHO it means we want some fields to be initialized as designated while other fields default initialized. It **explicitly** leave the 'not designated fields' to be default initialized. So`missing-field-initializers` should not report it as warning.

Our project builds failed with this change. We want the diagnostic flag to report for fields we are really forget to init instead of reporting designated initializers that DO means other fields to be default initialized.

The GCC community is still discussing this issue and has not drawn a conclusion yet ( for C++ ).

**Comment by shafik** [comment]
So we are hitting [dcl.init.aggr p5.2](https://eel.is/c++draft/dcl.init.aggr#5.2):

> Otherwise, if the element is not a reference, the element is copy-initialized from an empty initializer list ([dcl.init.list]).

and if we try another example w/o designated init: https://godbolt.org/z/Yn48KWvrM

```cpp
ExampleStruct buildExampleStruct(int a) {
    ExampleStruct e{1};
    return e;
}
```

We obtain the same diagnostic, so we are being entirely consistent here. I don't see a convincing argument that we should not warn for designated init but we should warn for non-designated aggregate init. The standard does not differentiate these cases.

**Comment by AaronBallman** [comment]
> IMHO it means we want some fields to be initialized as designated while other fields default initialized. It **explicitly** leave the 'not designated fields' to be default initialized. So`missing-field-initializers` should not report it as warning.

I think that any field initializer not explicitly mentioned is missing (per the usual English sense of the word) and to get your reading of the diagnostic, we'd want it named differently.

> Our project builds failed with this change. We want the diagnostic flag to report for fields we are really forget to init instead of reporting designated initializers that DO means other fields to be default initialized.

That said, this could be a reason to add a new group under `-Wmissing-field-initializers` for this functionality. I'm curious what the GCC folks think on the topic.

**Comment by gracicot** [comment]
@jamesruan Also note that with this warning, it's really easy for an implementer to specify when that warning happen.

```c++
struct AB {
    int a = {}; // A is an optional struct parameter
    int b;
};

AB test{
    .b = 0 // no warning for a since it has a default value
};
```

That way this warning only happen when users really make a mistake.

**Comment by floooh** [comment]
It looks to me like the missing-field-initializers warning for designated initialization in Clang 18.0.0 is also triggered in C, which definitely looks like a bug (since missing field initializers for designated init in C are a actually a very useful feature, such items are initialized to zero, can be detected and then patched to default values - FWIW, this change breaks all my C libraries).

The Clang 18 release notes say this:

- Clang now reports missing-field-initializers warning for missing designated initializers in C++. ([#56628](https://github.com/llvm/llvm-project/issues/56628))

...but it also happens in C mode.

I'm seeing this issue in the latest Emscripten SDK (3.1.51) which is using Clang 18.0.0

**Comment by shafik** [comment]
Note there are several discussions in the review: http://108.170.204.19/D157879

notably: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=96868

CC @Fznamznon @AaronBallman

**Comment by Fznamznon** [comment]
@floooh , the issue you're seeing can be a side effect of https://github.com/llvm/llvm-project/pull/70829 which I reverted before holidays season since there was no intention in enabling the warnings in C. Please check with clang built with trunk.

**Comment by marcushultman** [comment]
I seem to be unable to disable this with `-Wno-missing-field-initializers` in Android NDK r27c (based on clang-r522817c version 18.0.3).
